trinity/main.py

```diff
@@ -0,0 +1,249 @@
+        args.func(chain, use_ipython=use_ipython, debug=debug)
+        sys.exit(0)
+
+    logger, log_queue, listener = setup_trinity_logging(args.log_level.upper())
+
+    # start the listener thread to handle logs produced by other processes in
+    # the local logger.
+    listener.start()
+
+    # First initialize the database process.
+    database_server_process = ctx.Process(
+        target=run_database_process,
+        args=(
+            chain_config,
+            LevelDB,
+        ),
+        kwargs={'log_queue': log_queue}
+    )
+
+    # For now we just run the light sync against ropsten by default.
+    networking_process = ctx.Process(
+        target=run_networking_process,
+        args=(chain_config, sync_mode),
+        kwargs={'log_queue': log_queue}
+    )
+
+    # start the processes
+    database_server_process.start()
+    wait_for_ipc(chain_config.database_ipc_path)
+
+    networking_process.start()
+
+    try:
+        networking_process.join()
+    except KeyboardInterrupt:
+        logger.info('Keyboard Interrupt: Stopping')
+        kill_process_gracefully(networking_process)
+        logger.info('KILLED networking_process')
+        kill_process_gracefully(database_server_process)
+        logger.info('KILLED database_server_process')
+
+
+@with_queued_logging
+def run_database_process(chain_config, db_class):
+    db = db_class(db_path=chain_config.database_dir)
+
+    serve_chaindb(db, chain_config.database_ipc_path)
+
+
+@with_queued_logging
+def run_networking_process(chain_config, sync_mode):
+    class DBManager(BaseManager):
+        pass
+
+    DBManager.register('get_db', proxytype=DBProxy)
+    DBManager.register('get_chaindb', proxytype=ChainDBProxy)
+
+    manager = DBManager(address=chain_config.database_ipc_path)
+    manager.connect()
+
+    chaindb = manager.get_chaindb()
+
+    if not is_data_dir_initialized(chain_config):
+        # TODO: this will only work as is for chains with known genesis
+        # parameters.  Need to flesh out how genesis parameters for custom
+        # chains are defined and passed around.
+        initialize_data_dir(chain_config)
+
+    if not is_database_initialized(chaindb):
+        initialize_database(chain_config, chaindb)
+
+    chain_class = get_chain_protocol_class(chain_config, sync_mode=sync_mode)
+    peer_pool = PeerPool(LESPeer, chaindb, chain_config.network_id, chain_config.nodekey)
+
+    async def run():
+        try:
+            asyncio.ensure_future(peer_pool.run())
+            # chain.run() will run in a loop until our atexit handler is called, at which point it returns
+            # and we cleanly stop the pool and chain.
+            await chain.run()
+        finally:
+            await peer_pool.stop()
+            await chain.stop()
+
+    chain = chain_class(chaindb, peer_pool)
+
+    loop = asyncio.get_event_loop()
+
+    try:
+        loop.run_until_complete(run())
+    except KeyboardInterrupt:
```

Comment:
Will we ever hit this code path given that this is running in a sub-process? ISTM a KeyboardInterrupt would happen only in the main() function above. Also, instead of catching KeyboardInterrupt it'd be better to install a signal handler as that'd cover both a user hitting Ctrl-C in the terminal as well as the process being `kill`ed

Reply:
I just saw some of the signal handler stuff in one of your recent branches.  I'm going to open an issue to track this as handling keyboard interrupts is currently not working well and needs to be overhauled to use that mechanism.